### PR TITLE
Bug 1853311 - Add review checker reanalysis logic

### PR DESCRIPTION
**Pull request by rahulsainani**
### What
– Integrate `reanalyzeProduct` and `requestAnalysisStatus` APIs in `ReviewQualityCheckService` and middleware.
– Add polling logic for analysis status.
– Create a generic retry function.
– Add an edge case in mapper and test.

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1853311

### CODE REVIEW

File: fenix/app/src/test/java/org/mozilla/fenix/shopping/middleware/ProductAnalysisMapperTest.kt [hunk 1]
```diff
@@ -113,14 +113,29 @@ class ProductAnalysisMapperTest {
     }
 
     @Test
-    fun `WHEN product id is null THEN it is mapped to no analysis present`() {
```
Comment (calumozilla) [inline fenix/app/src/test/java/org/mozilla/fenix/shopping/middleware/ProductAnalysisMapperTest.kt:116]: do we also need tests to check AnalysisStatusDto and if the loading analysis card is present

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt [hunk 2]
```diff
@@ -34,15 +49,95 @@ class DefaultReviewQualityCheckService(
...
+    }
+
+    private inline fun <reified T : Enum<T>> String.asEnumOrDefault(defaultValue: T? = null): T? =
+        enumValues<T>().firstOrNull { it.name.equals(this, ignoreCase = true) } ?: defaultValue
+}
+
+/**
+ * Enum that represents the status of the product review analysis.
+ */
+enum class AnalysisStatusDto {
```
Comment (calumozilla) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt:113]: what does the `Dto` part of the name mean

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt [hunk 3]
```diff
@@ -46,21 +50,82 @@ class ReviewQualityCheckNetworkMiddleware(
...
+                    logger.info("Fetching product analysis")
+                    val productReviewState = fetchAnalysis()
+                    logger.debug("Product review state: $productReviewState")
+                    store.updateProductReviewState(productReviewState)
                 }
             }
 
             ReviewQualityCheckAction.ReanalyzeProduct -> {
-                // Bug 1853311 - Integrate analyze and analysis_status
+                scope.launch {
```
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt:69]: wondering aloud if the entire `when` should just be wrapped with `scope.launch{ }`

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt [hunk 4]
```diff
@@ -46,21 +50,82 @@ class ReviewQualityCheckNetworkMiddleware(
...
+                    val reanalysis = reviewQualityCheckService.reanalyzeProduct().also {
+                        logger.debug("Reanalysis: $it")
+                    }
+
+                    if (reanalysis == null) {
+                        store.updateProductReviewState(ProductReviewState.Error.GenericError)
+                        return@launch
```
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt:76]: what's the thought process of `return@launch` vs just `return` in this scenario? To ensure the coroutine terminates properly?

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt [hunk 5]
```diff
@@ -46,21 +50,82 @@ class ReviewQualityCheckNetworkMiddleware(
...
+                                store.updateProductReviewState(ProductReviewState.NoAnalysisPresent())
+                            } else if (state.productReviewState is ProductReviewState.AnalysisPresent) {
+                                store.updateProductReviewState(
+                                    state.productReviewState.copy(
+                                        analysisStatus = AnalysisStatus.NEEDS_ANALYSIS,
+                                    ),
+                                )
+                            }
+                        }
+                        return@launch
```
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt:98]: is this line needed?

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt [hunk 6]
```diff
@@ -46,21 +50,82 @@ class ReviewQualityCheckNetworkMiddleware(
...
+        retry(
+            predicate = { it == AnalysisStatusDto.PENDING || it == AnalysisStatusDto.IN_PROGRESS },
+            func = { reviewQualityCheckService.analysisStatus() },
+        ).also {
+            logger.debug("Status: $it")
+        }
+
+    private suspend fun AnalysisStatusDto.toProductReviewState(): ProductReviewState =
+        when (this) {
+            AnalysisStatusDto.COMPLETED -> fetchAnalysis()
```
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt:121]: can we maybe restructure this + `fetchAnalysis` slightly? they currently have a bit of a circular dependency 😨 

Maybe have the callers of `fetchAnalysis` call `toProductReviewState` instead, so the fetching is an isolated function and the transformation is a separate distinct step?

It feels off to have `toProductReviewState`, which feels like a state transformation/mapping function, firing off the network call, call itself, and return the result of calling itself back to `AnalysisStatusDto.COMPLETED`.

**Review by MozillaNoah — COMMENTED** [review]
first pass

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/Retry.kt [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.shopping.middleware
+
+import kotlinx.coroutines.delay
+
+private const val MAX_RETRIES = 10
+private const val INITIAL_DELAY_MS = 2000L
+private const val MAX_DELAY_MS = 20000L
+private const val FACTOR = 2.0
```
[the remark this answers is not in the document]
Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/Retry.kt:12]: Reanalysis takes 1-2 minutes and these delays work well for cases when the reanalysis is already wrapping up and then the initial delay is only 2 seconds, when a very patient user wants to see the whole loading, the max delay here would be 130 seconds, factor 2 being exponential backoff - double delay between each retry.

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt (hunk 2 above)
Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckService.kt:113]: Data transfer object, we are separate the `State` which is a UI model and the objects exposed via toolkit/GV/AC which are regarded as DTOs as they match the rest api one to one.

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/Retry.kt [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+ * Retry a suspend function until it returns a value that satisfies the predicate.
+ *
+ * @param maxRetries The maximum number of retries.
+ * @param initialDelayMs The initial delay in milliseconds.
+ * @param maxDelayMs The maximum delay in milliseconds.
+ * @param factor The factor to increase the delay by.
+ * @param predicate The predicate to check the result against.
+ * @param func The function to retry.
+ */
+suspend fun <T> retry(
```
[the remark this answers is not in the document]
Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/Retry.kt:24]: I will talk about it in one of the engineering syncs, this and `NetworkChecker` has the potential to be more common functionality across the project.

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt (hunk 5 above)
Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt:98]: No it is not, good catch!

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt (hunk 4 above)
Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt:76]: Since we're inside the a coroutine and this lamda/function is the `block` param of the `launch` function, we can only `return@launch`. The parent function has already executed the lines below `launch{}`,

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt (hunk 3 above)
Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt]: that's a good idea!

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt (hunk 6 above)
Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt:118]: They don't have a circular dependency – `fetchAnalysis` is calling `ProductAnalysis.toProductReviewState()` not this function. This function calls `fetchAnalysis`.

It's data logic that's acting on the response of one endpoint - deciding to fetch when the analysis is completed and returns errors in other cases. 

And yes, I think we agree that a mapper is a good place for functions that are just transformations, that's also another reason why this is not in `ProductAnalysisMapper`.

File: fenix/app/src/test/java/org/mozilla/fenix/shopping/middleware/ProductAnalysisMapperTest.kt (hunk 1 above)
Comment (rahulsainani) [reply fenix/app/src/test/java/org/mozilla/fenix/shopping/middleware/ProductAnalysisMapperTest.kt:116]: indeed, there are now tests in `ReviewQualityCheckStoreTest` for the analysis status. 

As for loading, I thought about it but it's not straightforward at this moment AFAIK. It's an intermediate state between the initial and the ending success/error states. It would involve observing the state in the test and asserting that there's loading in the middle. Not sure how to do that, let me know if you have any ideas. Also, depends on how pedantic we want to be about asserting the intermediate state.

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/Retry.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/Retry.kt:12]: Gotcha gotcha. makes sense to me

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt (hunk 6 above)
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNetworkMiddleware.kt:118]: Ahh gotcha gotcha, sorry for the confusion on my end then. This looks good 👍

**Review by MozillaNoah — APPROVED** [review]
LGTM! Tested locally with a product listing for a microphone and it had an out of date review, switched to the analysis in progress UI, and dismissed the UI and updated the rating.
